### PR TITLE
[SPARK-42584][CONNECT] Improve output of `Column.explain`

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, connect display the structure of the proto in both the regular and extended version of explain. We should display a more compact sql-a-like string for the regular version.

### Why are the changes needed?
Improve output of Column.explain so as display the same as the explain API of sql expression.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @hvanhovell cc @HyukjinKwon @zhengruifeng @amaliujia @ueshin @LuciferYang

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto [hunk 1]
```diff
@@ -163,6 +164,14 @@ message AnalyzePlanRequest {
     // (Required) The logical plan to get a hashCode.
     Plan plan = 1;
   }
+
+  // Explains the expression based on extended is true or not.
+  message ExplainExpression {
+    // (Required) The expression to be analyzed.
+    Expression expr = 1;
+
+    bool extended = 2;
```
Comment (amaliujia) [inline connector/connect/common/src/main/protobuf/spark/connect/base.proto:173]: Nit: please follow the style guide to indicate if this is required or optional and also document the semantics.

**Comment by amaliujia** [comment]
Do we need python side compatible API?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala [hunk 2]
```diff
@@ -1211,13 +1211,11 @@ class Column private[sql] (private[sql] val expr: proto.Expression) extends Logg
    * @group df_ops
    * @since 3.4.0
    */
-  def explain(extended: Boolean): Unit = {
+  def explain(extended: Boolean)(implicit spark: SparkSession): Unit = {
```
Comment (ueshin) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1214]: Who will provide the implicit `SparkSession` in the users' context?

Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1214]: Good question. It seems we must send the msg to server side. So we require `SparkSession`.
But it's a question how to get `SparkSession` in connect's `Column` ? or we define the SparkSession in `Column` directly?

**Comment by beliefer** [comment]
> Do we need python side compatible API?

Maybe. I will check the python side.

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto (hunk 1 above)
Comment (beliefer) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:173]: Thank you! I forgot it.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala (hunk 2 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1214]: I am not sure what happened here, my comment seemed to have gone missing... Anyway, please don't send an RPC for this. Just use the local expression tree and create something that is a little bit easier to read.

Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1214]: @hvanhovell If so, the output of `Column.explain` can't keep consistent with Spark SQL.

Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1214]: @hvanhovell Do you mean to construct local catalyst expression tree at connect client side ?

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1214]: @beliefer no we can't be fully consistent with SQL. We are already deviating from what the original implementation does when you set `extended = true`. I just want something that is more readable.

As for the implementation. The catalyst dependency is going away. Just create a bit of tree traversal logic that nicely renders the expression tree.

Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:1214]: I understand what you mean roughly. We don't need follow SQL, but the output reflects the hierarchy of the tree.

**Comment by beliefer** [comment]
https://github.com/apache/spark/pull/40528 used to replace this one.

**Pull request closed without merging**